src/controller_examples/zookeeper_controller/proof/safety.rs

```diff
@@ -392,6 +395,1027 @@ pub open spec fn sts_update_request_msg_since(key: ObjectRef, rest_id: RestId) -
+                    && resp_msg_matches_req_msg(resp_msg, s.pending_req_of(key))
+                })
+            }
+    }
+}
+
+pub proof fn lemma_always_pending_req_in_flight_or_resp_in_flight_at_after_update_stateful_set_step(
+    spec: TempPred<ClusterState>, key: ObjectRef
+)
+    requires
+        spec.entails(lift_state(init::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>())),
+        spec.entails(always(lift_action(next::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>()))),
+        spec.entails(always(lift_state(each_resp_matches_at_most_one_pending_req::<ZookeeperClusterView, ZookeeperReconcileState>(key)))),
+    ensures
+        spec.entails(
+            always(lift_state(pending_req_in_flight_or_resp_in_flight_at_after_update_stateful_set_step(key)))
+        ),
+{
+    let invariant = pending_req_in_flight_or_resp_in_flight_at_after_update_stateful_set_step(key);
+    let stronger_next = |s, s_prime: ClusterState| {
+        &&& next::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>()(s, s_prime)
+        &&& each_resp_matches_at_most_one_pending_req::<ZookeeperClusterView, ZookeeperReconcileState>(key)(s)
+    };
+    assert forall |s, s_prime: ClusterState| invariant(s) && #[trigger] stronger_next(s, s_prime) implies invariant(s_prime) by {
+        if at_after_update_stateful_set_step(key)(s_prime) {
+            let step = choose |step| next_step::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>(s, s_prime, step);
+            assert(!step.is_RestartController());
+            let resp = choose |msg| {
+                #[trigger] s.message_in_flight(msg)
+                && resp_msg_matches_req_msg(msg, s.pending_req_of(key))
+            };
+            match step {
+                Step::KubernetesAPIStep(input) => {
+                    if input == Option::Some(s.pending_req_of(key)) {
+                        let resp_msg = handle_update_request(s.pending_req_of(key), s.kubernetes_api_state).1;
+                        assert(s_prime.message_in_flight(resp_msg));
+                    } else {
+                        if !s.message_in_flight(s.pending_req_of(key)) {
+                            assert(s_prime.message_in_flight(resp));
+                        }
+                    }
+                }
+                Step::ControllerStep(input) => {
+                    let cr_key = input.1.get_Some_0();
+                    if cr_key != key {
+                        if s.message_in_flight(s.pending_req_of(key)) {
+                            assert(s_prime.message_in_flight(s_prime.pending_req_of(key)));
+                        } else {
+                            assert(s_prime.message_in_flight(resp));
+                        }
+                    } else {
+                        assert(s_prime.message_in_flight(s_prime.pending_req_of(key)));
+                    }
+                }
+                Step::ClientStep(input) => {
```

Comment:
Would the proof still go through without reasoning about ClientStep (similarly for the lemmas below)?

Reply:
No, I tried without it at first.